### PR TITLE
Code for "Gradient Harmonized Single-stage Detector"

### PR DESCRIPTION
**Pull request by libuyu**

### CODE REVIEW

File: mmdet/models/losses/ghm_loss.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    return bin_labels, bin_label_weights
+
+
+@LOSSES.register_module
+class GHMC(nn.Module):
+    def __init__(
+            self,
+            bins=10,
+            momentum=0,
+            use_sigmoid=True,
```
Comment (hellock) [inline mmdet/models/losses/ghm_loss.py]: If there is no plan to support softmax, this argument can be removed.

File: mmdet/models/losses/ghm_loss.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        self.momentum = momentum
+        self.edges = [float(x) / bins for x in range(bins+1)]
+        self.edges[-1] += 1e-6
+        if momentum > 0:
+            self.acc_sum = [0.0 for _ in range(bins)]
+        self.use_sigmoid = use_sigmoid
+        self.loss_weight = loss_weight
+
+    def forward(self, pred, target, label_weight, *args, **kwargs):
+        """ Args:
```
Comment (hellock) [inline mmdet/models/losses/ghm_loss.py]: Please refer to https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html for the docstring style.

File: mmdet/models/losses/ghm_loss.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    inds = torch.nonzero(labels >= 1).squeeze()
+    if inds.numel() > 0:
+        bin_labels[inds, labels[inds] - 1] = 1
+    bin_label_weights = label_weights.view(-1, 1).expand(
+        label_weights.size(0), label_channels)
+    return bin_labels, bin_label_weights
+
+
+@LOSSES.register_module
+class GHMC(nn.Module):
```
Comment (hellock) [inline mmdet/models/losses/ghm_loss.py]: Add a docstring for the class.

File: mmdet/models/losses/ghm_loss.py [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        if n > 0:
+            weights = weights / n
+
+        loss = F.binary_cross_entropy_with_logits(
+            pred, target, weights, reduction='sum') / tot
+        return loss * self.loss_weight
+
+
+@LOSSES.register_module
+class GHMR(nn.Module):
```
Comment (hellock) [inline mmdet/models/losses/ghm_loss.py]: Add a docstring for the class.

File: mmdet/models/losses/ghm_loss.py [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    def __init__(
+            self,
+            bins=10,
+            momentum=0,
+            use_sigmoid=True,
+            loss_weight=1.0):
+        super(GHMC, self).__init__()
+        self.bins = bins
+        self.momentum = momentum
+        self.edges = [float(x) / bins for x in range(bins+1)]
```
Comment (hellock) [inline mmdet/models/losses/ghm_loss.py]: May use `np.arange` to simplify this.

File: mmdet/models/losses/ghm_loss.py (hunk 1 above)
Comment (libuyu) [reply mmdet/models/losses/ghm_loss.py]: https://github.com/libuyu/mmdetection/blob/143a1ec5ff42eecf6576252ae91dd5ddeb8de70e/mmdet/models/anchor_heads/anchor_head.py#L59
if "us_sigmoid" removed, this line will be False and an error occurs.

File: mmdet/models/losses/ghm_loss.py (hunk 5 above)
Comment (libuyu) [reply mmdet/models/losses/ghm_loss.py]: We prefer a tensorized version of tools where numpy is not used for acceleration. So I think torch.arange may be better.

**Comment by libuyu** [comment]
Revised

File: mmdet/models/losses/ghm_loss.py (hunk 5 above)
Comment (hellock) [reply mmdet/models/losses/ghm_loss.py]: Either np or torch is ok to me since edges are not involved in any tensor ops and only used as scalars. Using cuda tensors here will not be faster.

File: mmdet/models/losses/ghm_loss.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+            pred (float tensor of size [batch_num, class_num]):
+                The direct prediction of classification fc layer.
+            target (float tensor of size [batch_num, class_num]):
+                Binary class target for each sample.
+            label_weight (float tensor of size [batch_num, class_num]):
+                the value is 1 if the sample is valid and 0 if ignored.
+
+        Returns:
+            The gradient harmonized loss.
+
```
Comment (hellock) [inline mmdet/models/losses/ghm_loss.py]: This empty line is unnecessary.

File: mmdet/models/losses/ghm_loss.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        """
+        if not self.use_sigmoid:
+            raise NotImplementedError
```
Comment (hellock) [inline mmdet/models/losses/ghm_loss.py]: This assertion can be moved to the init method to have an early check.

File: mmdet/models/losses/ghm_loss.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        label_weights.size(0), label_channels)
+    return bin_labels, bin_label_weights
+
+
+@LOSSES.register_module
+class GHMC(nn.Module):
+    """GHM Classification Loss.
+
+    Details of the theorem can be viewed in the paper
+    "Gradient Harmonized Single-stage Detector"
```
Comment (hellock) [inline mmdet/models/losses/ghm_loss.py]: May also provide an arxiv url here for quick reference.

Docstring for classes will document the args of `__init__` method.

File: mmdet/models/losses/ghm_loss.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+            pred (float tensor of size [batch_num, 4 (* class_num)]):
+                The prediction of box regression layer. Channel number can be 4
+                or 4 * class_num depending on whether it is class-agnostic.
+            target (float tensor of size [batch_num, 4 (* class_num)]):
+                The target regression values with the same size of pred.
+            label_weight (float tensor of size [batch_num, 4 (* class_num)]):
+                The weight of each sample, 0 if ignored.
+        Returns:
+            The gradient harmonized loss.
+
```
Comment (hellock) [inline mmdet/models/losses/ghm_loss.py]: This empty line is unnecessary.

**Comment by libuyu** [comment]
All revised now.

**Comment by OpenMMLab-Assistant001** [comment]
Hi @libuyu ！

First of all, we want to express our gratitude for your significant PR in the mmdetection project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）

Thank you again for your contribution❤
